// @arizeai/openinference-core/dist/src/trace/contextAttributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextAttributes = void 0;
exports.setPromptTemplate = setPromptTemplate;
exports.clearPromptTemplate = clearPromptTemplate;
exports.getPromptTemplate = getPromptTemplate;
exports.setSession = setSession;
exports.clearSession = clearSession;
exports.getSession = getSession;
exports.setMetadata = setMetadata;
exports.clearMetadata = clearMetadata;
exports.getMetadata = getMetadata;
exports.setUser = setUser;
exports.clearUser = clearUser;
exports.getUser = getUser;
exports.setTags = setTags;
exports.clearTags = clearTags;
exports.getTags = getTags;
exports.setAttributes = setAttributes;
exports.clearAttributes = clearAttributes;
exports.getAttributes = getAttributes;
exports.getAttributesFromContext = getAttributesFromContext;
const api_1 = require("@opentelemetry/api");
const core_1 = require("@opentelemetry/core");
const openinference_semantic_conventions_1 = require("@arizeai/openinference-semantic-conventions");
const utils_1 = require("../utils");
const CONTEXT_ATTRIBUTES_ATTRIBUTES_KEY = "attributes";
exports.ContextAttributes = {
    [openinference_semantic_conventions_1.PROMPT_TEMPLATE_TEMPLATE]: (0, api_1.createContextKey)(`OpenInference SDK Context Key ${openinference_semantic_conventions_1.PROMPT_TEMPLATE_TEMPLATE}`),
    [openinference_semantic_conventions_1.PROMPT_TEMPLATE_VARIABLES]: (0, api_1.createContextKey)(`OpenInference SDK Context Key ${openinference_semantic_conventions_1.PROMPT_TEMPLATE_VARIABLES}`),
    [openinference_semantic_conventions_1.PROMPT_TEMPLATE_VERSION]: (0, api_1.createContextKey)(`OpenInference SDK Context Key ${openinference_semantic_conventions_1.PROMPT_TEMPLATE_VERSION}`),
    [openinference_semantic_conventions_1.SESSION_ID]: (0, api_1.createContextKey)(`OpenInference SDK Context Key ${openinference_semantic_conventions_1.SESSION_ID}`),
    [openinference_semantic_conventions_1.METADATA]: (0, api_1.createContextKey)(`OpenInference SDK Context Key ${openinference_semantic_conventions_1.METADATA}`),
    [openinference_semantic_conventions_1.USER_ID]: (0, api_1.createContextKey)(`OpenInference SDK Context Key ${openinference_semantic_conventions_1.USER_ID}`),
    [openinference_semantic_conventions_1.TAG_TAGS]: (0, api_1.createContextKey)(`OpenInference SDK Context Key ${openinference_semantic_conventions_1.TAG_TAGS}`),
    [CONTEXT_ATTRIBUTES_ATTRIBUTES_KEY]: (0, api_1.createContextKey)(`OpenInference SDK Context Key attributes`),
};
const { [openinference_semantic_conventions_1.PROMPT_TEMPLATE_TEMPLATE]: PROMPT_TEMPLATE_TEMPLATE_KEY, [openinference_semantic_conventions_1.PROMPT_TEMPLATE_VARIABLES]: PROMPT_TEMPLATE_VARIABLES_KEY, [openinference_semantic_conventions_1.PROMPT_TEMPLATE_VERSION]: PROMPT_TEMPLATE_VERSION_KEY, [openinference_semantic_conventions_1.SESSION_ID]: SESSION_ID_KEY, [openinference_semantic_conventions_1.METADATA]: METADATA_KEY, [openinference_semantic_conventions_1.USER_ID]: USER_ID_KEY, [openinference_semantic_conventions_1.TAG_TAGS]: TAG_TAGS_KEY, [CONTEXT_ATTRIBUTES_ATTRIBUTES_KEY]: ATTRIBUTES_KEY, } = exports.ContextAttributes;
/**
 * Sets the prompt template in the given OpenTelemetry context.
 * If OI tracer is used in spans created during this active context, the prompt template will be added to the span as an attribute.
 * @param context - The context to set the prompt template on.
 * @param promptTemplate - The prompt template to set on the context.
 * @returns The context
 * @example
 * ```typescript
 * context.with(
 *   setPromptTemplate(context.active(), {
 *     template: "hello {name}",
 *     variables: { name: "world" },
 *     version: "V1.0"
 *   }),
 *   () => {
 *     // Spans created here will have prompt template attributes
 *   }
 * );
 * ```
 */
function setPromptTemplate(context, promptTemplate) {
    const { template, variables, version } = promptTemplate;
    context = context.setValue(PROMPT_TEMPLATE_TEMPLATE_KEY, template);
    if (variables) {
        context = context.setValue(PROMPT_TEMPLATE_VARIABLES_KEY, (0, utils_1.safelyJSONStringify)(variables));
    }
    if (version) {
        context = context.setValue(PROMPT_TEMPLATE_VERSION_KEY, version);
    }
    return context;
}
/**
 * Clears the prompt template from the given OpenTelemetry context.
 * @param context - The context to clear the prompt template from.
 * @returns The context
 */
function clearPromptTemplate(context) {
    context = context.deleteValue(PROMPT_TEMPLATE_TEMPLATE_KEY);
    context = context.deleteValue(PROMPT_TEMPLATE_VARIABLES_KEY);
    context = context.deleteValue(PROMPT_TEMPLATE_VERSION_KEY);
    return context;
}
/**
 * Gets the prompt template from the given OpenTelemetry context.
 * @param context - The context to get the prompt template from.
 * @returns The prompt template
 * @example const promptTemplate = getPromptTemplate(context.active());
 */
function getPromptTemplate(context) {
    const maybeTemplate = context.getValue(PROMPT_TEMPLATE_TEMPLATE_KEY);
    const maybeVariables = context.getValue(PROMPT_TEMPLATE_VARIABLES_KEY);
    const maybeVersion = context.getValue(PROMPT_TEMPLATE_VERSION_KEY);
    const attributes = {};
    if (typeof maybeTemplate === "string") {
        attributes.template = maybeTemplate;
    }
    if (typeof maybeVariables === "string") {
        const parsedVariables = (0, utils_1.safelyJSONParse)(maybeVariables);
        attributes.variables = (0, utils_1.isObjectWithStringKeys)(parsedVariables) ? parsedVariables : undefined;
    }
    if (typeof maybeVersion === "string") {
        attributes.version = maybeVersion;
    }
    if (Object.keys(attributes).length === 0) {
        return;
    }
    return attributes;
}
/**
 * Sets the session ID in the given OpenTelemetry context.
 * If OI tracer is used in spans created during this active context, the session ID will be added to the span as an attribute.
 * @param context - The context to set the session ID on.
 * @param session - The session to set on the context.
 * @returns The context
 * @example
 * ```typescript
 * context.with(
 *   setSession(context.active(), { sessionId: "session-123" }),
 *   () => {
 *     // Spans created here will have session ID attribute
 *   }
 * );
 * ```
 */
function setSession(context, session) {
    const { sessionId } = session;
    return context.setValue(SESSION_ID_KEY, sessionId);
}
function clearSession(context) {
    return context.deleteValue(SESSION_ID_KEY);
}
/**
 * Retrieves the session ID from the given context.
 * @param context - The context object.
 * @returns {string | undefined} The session ID if it exists, otherwise undefined.
 */
function getSession(context) {
    const maybeSessionId = context.getValue(SESSION_ID_KEY);
    if (typeof maybeSessionId === "string") {
        return { sessionId: maybeSessionId };
    }
}
/**
 * Sets the metadata in the given OpenTelemetry context.
 * If OI tracer is used in spans created during this active context, the metadata will be added to the span as an attribute.
 * @param context - The context to set the metadata on.
 * @param metadata - The metadata to set on the context.
 * @returns The context
 * @example
 * ```typescript
 * context.with(
 *   setMetadata(context.active(), { key: "value", numeric: 1 }),
 *   () => {
 *     // Spans created here will have metadata attribute
 *   }
 * );
 * ```
 */
function setMetadata(context, metadata) {
    return context.setValue(METADATA_KEY, (0, utils_1.safelyJSONStringify)(metadata));
}
/**
 * Clears the metadata from the given OpenTelemetry context.
 * @param context - The context to clear the metadata from.
 * @returns The context
 */
function clearMetadata(context) {
    return context.deleteValue(METADATA_KEY);
}
/**
 * Gets the metadata from the given OpenTelemetry context.
 * @param context - The context to get the metadata from.
 * @returns The metadata
 * @example const metadata = getMetadata(context.active());
 */
function getMetadata(context) {
    const maybeMetadata = context.getValue(METADATA_KEY);
    if (typeof maybeMetadata === "string") {
        const parsedMetadata = (0, utils_1.safelyJSONParse)(maybeMetadata);
        return (0, utils_1.isObjectWithStringKeys)(parsedMetadata) ? parsedMetadata : undefined;
    }
}
/**
 * Sets the user ID in the given OpenTelemetry context.
 * @param context - The context to set the user ID on.
 * @param user - The user to set on the context.
 * @returns The context
 * @example
 * ```typescript
 * context.with(
 *   setUser(context.active(), { userId: "user-123" }),
 *   () => {
 *     // Spans created here will have user ID attribute
 *   }
 * );
 * ```
 */
function setUser(context, user) {
    const { userId } = user;
    return context.setValue(USER_ID_KEY, userId);
}
/**
 * Clears the user ID from the given OpenTelemetry context.
 * @param context - The context to clear the user ID from.
 * @returns The context
 */
function clearUser(context) {
    return context.deleteValue(USER_ID_KEY);
}
/**
 * Gets the user ID from the given OpenTelemetry context.
 * @param context - The context to get the user ID from.
 * @returns The user
 * @example const user = getUser(context.active());
 */
function getUser(context) {
    const maybeUserId = context.getValue(USER_ID_KEY);
    if (typeof maybeUserId === "string") {
        return { userId: maybeUserId };
    }
}
/**
 * Sets the tags in the given OpenTelemetry context.
 * If OI tracer is used in spans created during this active context, the tags will be added to the span as an attribute.
 * @param context - The context to set the tags on.
 * @param tags - The tags to set on the context.
 * @returns The context
 * @example
 * ```typescript
 * context.with(
 *   setTags(context.active(), ["tag1", "tag2"]),
 *   () => {
 *     // Spans created here will have tags attribute
 *   }
 * );
 * ```
 */
function setTags(context, tags) {
    return context.setValue(TAG_TAGS_KEY, (0, utils_1.safelyJSONStringify)(tags));
}
/**
 * Clears the tags from the given OpenTelemetry context.
 * @param context - The context to clear the tags from.
 * @returns The context
 */
function clearTags(context) {
    return context.deleteValue(TAG_TAGS_KEY);
}
/**
 * Gets the tags from the given OpenTelemetry context.
 * @param context - The context to get the tags from.
 * @returns The tags
 * @example const tags = getTags(context.active());
 */
function getTags(context) {
    const maybeTags = context.getValue(TAG_TAGS_KEY);
    if (typeof maybeTags === "string") {
        const parsedTags = (0, utils_1.safelyJSONParse)(maybeTags);
        return (0, utils_1.isStringArray)(parsedTags) ? parsedTags : undefined;
    }
}
/**
 * Sets the attributes in the given OpenTelemetry context.
 * If OI tracer is used in spans created during this active context, the attributes will be added to the span as an attribute.
 * @param context - The context to set the attributes on.
 * @param attributes - The attributes to set on the context.
 * @returns The context
 * @example
 * ```typescript
 * context.with(
 *   setAttributes(context.active(), { custom: "value", test: "attribute" }),
 *   () => {
 *     // Spans created here will have custom attributes
 *   }
 * );
 * ```
 */
function setAttributes(context, attributes) {
    return context.setValue(ATTRIBUTES_KEY, (0, utils_1.safelyJSONStringify)(attributes));
}
/**
 * Clears the attributes from the given OpenTelemetry context.
 * @param context - The context to clear the attributes from.
 * @returns The context
 */
function clearAttributes(context) {
    return context.deleteValue(ATTRIBUTES_KEY);
}
/**
 * Gets the attributes from the given OpenTelemetry context.
 * @param context - The context to get the attributes from.
 * @returns The attributes
 * @example const attributes = getAttributes(context.active());
 */
function getAttributes(context) {
    const maybeAttributes = context.getValue(ATTRIBUTES_KEY);
    if (typeof maybeAttributes === "string") {
        const parsedAttributes = (0, utils_1.safelyJSONParse)(maybeAttributes);
        return (0, utils_1.isAttributes)(parsedAttributes) ? parsedAttributes : undefined;
    }
}
/**
 * Gets the OpenInference attributes from the given context
 * @param context
 * @example
 * ```typescript
 * // Chaining multiple attributes
 * context.with(
 *   setSession(
 *     setPromptTemplate(context.active(), {
 *       template: "hello {name}",
 *       variables: { name: "world" },
 *       version: "V1.0"
 *     }),
 *     { sessionId: "session-123" }
 *   ),
 *   () => {
 *     // Extract all attributes for span
 *     const attributes = getAttributesFromContext(context.active());
 *     span.setAttributes(attributes);
 *   }
 * );
 * ```
 * @returns {Attributes} The OpenInference attributes formatted as OpenTelemetry span attributes.
 */
function getAttributesFromContext(context) {
    let attributes = {};
    Object.entries(exports.ContextAttributes).forEach(([key, symbol]) => {
        const maybeValue = context.getValue(symbol);
        if (key === CONTEXT_ATTRIBUTES_ATTRIBUTES_KEY) {
            if (typeof maybeValue === "string") {
                const parsedAttributes = (0, utils_1.safelyJSONParse)(maybeValue);
                if ((0, utils_1.isAttributes)(parsedAttributes)) {
                    attributes = Object.assign(Object.assign({}, attributes), parsedAttributes);
                }
            }
            return;
        }
        if ((0, core_1.isAttributeValue)(maybeValue) && maybeValue !== undefined) {
            attributes[key] = maybeValue;
        }
    });
    return attributes;
}
