// @arizeai/openinference-instrumentation-mcp/dist/src/mcp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPInstrumentation = void 0;
exports.isPatched = isPatched;
const api_1 = require("@opentelemetry/api");
const instrumentation_1 = require("@opentelemetry/instrumentation");
const version_1 = require("./version");
const CLIENT_SSE_MODULE_NAME = "@modelcontextprotocol/sdk/client/sse";
const SERVER_SSE_MODULE_NAME = "@modelcontextprotocol/sdk/server/sse";
const CLIENT_STDIO_MODULE_NAME = "@modelcontextprotocol/sdk/client/stdio";
const SERVER_STDIO_MODULE_NAME = "@modelcontextprotocol/sdk/server/stdio";
const CLIENT_STREAMABLE_HTTP_MODULE_NAME = "@modelcontextprotocol/sdk/client/streamableHttp";
const SERVER_STREAMABLE_HTTP_MODULE_NAME = "@modelcontextprotocol/sdk/server/streamableHttp";
/**
 * Flags to check if a module has already been patched.
 * Note: This is a fallback in case the module is made immutable (e.x. Deno, webpack, etc.)
 */
const patchedModules = {
    [CLIENT_SSE_MODULE_NAME]: false,
    [SERVER_SSE_MODULE_NAME]: false,
    [CLIENT_STDIO_MODULE_NAME]: false,
    [SERVER_STDIO_MODULE_NAME]: false,
    [CLIENT_STREAMABLE_HTTP_MODULE_NAME]: false,
    [SERVER_STREAMABLE_HTTP_MODULE_NAME]: false,
};
/**
 * function to check if instrumentation is enabled / disabled for the module
 */
function isPatched(moduleName) {
    return patchedModules[moduleName];
}
/**
 * Instrumentation for the MCP SDK which propagates context between client and server to allow
 * traces to be connected across tool calls.
 * @param instrumentationConfig The config for the instrumentation @see {@link InstrumentationConfig}
 */
class MCPInstrumentation extends instrumentation_1.InstrumentationBase {
    constructor({ instrumentationConfig, } = {}) {
        super("@arizeai/openinference-instrumentation-mcp", version_1.VERSION, instrumentationConfig);
    }
    /**
     * Manually instruments the MCP client and/or server modules. Currently, auto-instrumentation does not work
     * with the MCP SDK and this method must be used to enable instrumentation.
     * @param {Object} modules - The modules to manually instrument.
     * @param {typeof ClientSSEModule} modules.clientSSEModule - The MCP client SSE module, e.g. require('@modelcontextprotocol/sdk/client/sse.js')
     * @param {typeof ServerSSEModule} modules.serverSSEModule - The MCP server SSE module, e.g. require('@modelcontextprotocol/sdk/server/sse.js')
     * @param {typeof ClientStdioModule} modules.clientStdioModule - The MCP client stdio module, e.g. require('@modelcontextprotocol/sdk/server/stdio.js')
     * @param {typeof ServerStdioModule} modules.serverStdioModule - The MCP server stdio module, e.g. require('@modelcontextprotocol/sdk/client/stdio.js')
     * @param {typeof ClientStreamableHTTPModule} modules.clientStreamableHTTPModule - The MCP client streamable HTTP module, e.g. require('@modelcontextprotocol/sdk/client/streamableHttp.js')
     * @param {typeof ServerStreamableHTTPModule} modules.serverStreamableHTTPModule - The MCP server streamable HTTP module, e.g. require('@modelcontextprotocol/sdk/server/streamableHttp.js')
     */
    manuallyInstrument({ clientSSEModule, serverSSEModule, clientStdioModule, serverStdioModule, clientStreamableHTTPModule, serverStreamableHTTPModule, }) {
        if (clientSSEModule) {
            api_1.diag.debug(`Manually instrumenting ${CLIENT_SSE_MODULE_NAME}`);
            this._patchClientSSEModule(clientSSEModule);
        }
        if (serverSSEModule) {
            api_1.diag.debug(`Manually instrumenting ${SERVER_SSE_MODULE_NAME}`);
            this._patchServerSSEModule(serverSSEModule);
        }
        if (clientStdioModule) {
            api_1.diag.debug(`Manually instrumenting ${CLIENT_STDIO_MODULE_NAME}`);
            this._patchClientStdioModule(clientStdioModule);
        }
        if (serverStdioModule) {
            api_1.diag.debug(`Manually instrumenting ${SERVER_STDIO_MODULE_NAME}`);
            this._patchServerStdioModule(serverStdioModule);
        }
        if (clientStreamableHTTPModule) {
            api_1.diag.debug(`Manually instrumenting ${CLIENT_STREAMABLE_HTTP_MODULE_NAME}`);
            this._patchClientStreamableHTTPModule(clientStreamableHTTPModule);
        }
        if (serverStreamableHTTPModule) {
            api_1.diag.debug(`Manually instrumenting ${SERVER_STREAMABLE_HTTP_MODULE_NAME}`);
            this._patchServerStreamableHTTPModule(serverStreamableHTTPModule);
        }
    }
    init() {
        return [
            new instrumentation_1.InstrumentationNodeModuleDefinition("@modelcontextprotocol/sdk/client/sse.js", [">=1.0.0"], this._patchClientSSEModule.bind(this), this._unpatchClientSSEModule.bind(this)),
            new instrumentation_1.InstrumentationNodeModuleDefinition("@modelcontextprotocol/sdk/server/sse.js", [">=1.0.0"], this._patchServerSSEModule.bind(this), this._unpatchServerSSEModule.bind(this)),
            new instrumentation_1.InstrumentationNodeModuleDefinition("@modelcontextprotocol/sdk/client/stdio.js", [">=1.0.0"], this._patchClientStdioModule.bind(this), this._unpatchClientStdioModule.bind(this)),
            new instrumentation_1.InstrumentationNodeModuleDefinition("@modelcontextprotocol/sdk/server/stdio.js", [">=1.0.0"], this._patchServerStdioModule.bind(this), this._unpatchServerStdioModule.bind(this)),
            new instrumentation_1.InstrumentationNodeModuleDefinition("@modelcontextprotocol/sdk/client/streamableHttp.js", [">=1.10.0"], this._patchClientStreamableHTTPModule.bind(this), this._unpatchClientStreamableHTTPModule.bind(this)),
            new instrumentation_1.InstrumentationNodeModuleDefinition("@modelcontextprotocol/sdk/server/streamableHttp.js", [">=1.10.0"], this._patchServerStreamableHTTPModule.bind(this), this._unpatchServerStreamableHTTPModule.bind(this)),
        ];
    }
    _patchClientSSEModule(module, moduleVersion) {
        return this._patchTransport(module, CLIENT_SSE_MODULE_NAME, moduleVersion, module.SSEClientTransport);
    }
    _unpatchClientSSEModule(module) {
        return this._unpatchTransport(module, CLIENT_SSE_MODULE_NAME, module.SSEClientTransport);
    }
    _patchServerSSEModule(module, moduleVersion) {
        return this._patchTransport(module, SERVER_SSE_MODULE_NAME, moduleVersion, module.SSEServerTransport);
    }
    _unpatchServerSSEModule(module) {
        return this._unpatchTransport(module, SERVER_SSE_MODULE_NAME, module.SSEServerTransport);
    }
    _patchClientStdioModule(module, moduleVersion) {
        return this._patchTransport(module, CLIENT_STDIO_MODULE_NAME, moduleVersion, module.StdioClientTransport);
    }
    _unpatchClientStdioModule(module) {
        return this._unpatchTransport(module, CLIENT_STDIO_MODULE_NAME, module.StdioClientTransport);
    }
    _patchServerStdioModule(module, moduleVersion) {
        return this._patchTransport(module, SERVER_STDIO_MODULE_NAME, moduleVersion, module.StdioServerTransport);
    }
    _unpatchServerStdioModule(module) {
        return this._unpatchTransport(module, SERVER_STDIO_MODULE_NAME, module.StdioServerTransport);
    }
    _patchClientStreamableHTTPModule(module, moduleVersion) {
        return this._patchTransport(module, CLIENT_STREAMABLE_HTTP_MODULE_NAME, moduleVersion, module.StreamableHTTPClientTransport);
    }
    _unpatchClientStreamableHTTPModule(module) {
        return this._unpatchTransport(module, CLIENT_STREAMABLE_HTTP_MODULE_NAME, module.StreamableHTTPClientTransport);
    }
    _patchServerStreamableHTTPModule(module, moduleVersion) {
        return this._patchTransport(module, SERVER_STREAMABLE_HTTP_MODULE_NAME, moduleVersion, module.StreamableHTTPServerTransport);
    }
    _unpatchServerStreamableHTTPModule(module) {
        return this._unpatchTransport(module, SERVER_STREAMABLE_HTTP_MODULE_NAME, module.StreamableHTTPServerTransport);
    }
    _patchTransport(module, moduleName, moduleVersion, transportClass) {
        api_1.diag.debug(`Applying patch for ${moduleName}@${moduleVersion}`);
        if ((module === null || module === void 0 ? void 0 : module.openInferencePatched) || patchedModules[moduleName]) {
            return module;
        }
        this._wrap(transportClass.prototype, "send", this._getTransportSendPatch());
        this._wrap(transportClass.prototype, "start", this._getTransportStartPatch());
        patchedModules[moduleName] = true;
        try {
            // This can fail if the module is made immutable via the runtime or bundler
            module.openInferencePatched = true;
        }
        catch (e) {
            api_1.diag.debug(`Failed to set ${moduleName} patched flag on the module`, e);
        }
        return module;
    }
    _unpatchTransport(module, moduleName, transportClass) {
        this._unwrap(transportClass.prototype, "send");
        this._unwrap(transportClass.prototype, "start");
        patchedModules[moduleName] = false;
        try {
            // This can fail if the module is made immutable via the runtime or bundler
            module.openInferencePatched = false;
        }
        catch (e) {
            api_1.diag.warn(`Failed to unset ${moduleName} patched flag on the module`, e);
        }
        return module;
    }
    _getTransportSendPatch() {
        return (original) => {
            return function send(...args) {
                const message = args[0];
                if (!MCPInstrumentation._isJSONRPCRequest(message)) {
                    return original.apply(this, args);
                }
                if (!message.params) {
                    message.params = {};
                }
                if (!message.params._meta) {
                    message.params._meta = {};
                }
                api_1.propagation.inject(api_1.context.active(), message.params._meta);
                return original.apply(this, args);
            };
        };
    }
    _getTransportStartPatch() {
        return (original) => {
            return function start(...args) {
                const onmessage = this.onmessage;
                if (!onmessage) {
                    return original.apply(this, args);
                }
                this.onmessage = (...args) => {
                    var _a;
                    const message = args[0];
                    if (!MCPInstrumentation._isJSONRPCRequest(message)) {
                        return onmessage.apply(this, args);
                    }
                    const ctx = api_1.propagation.extract(api_1.context.active(), (_a = message.params) === null || _a === void 0 ? void 0 : _a._meta);
                    return api_1.context.with(ctx, () => {
                        return onmessage.apply(this, args);
                    });
                };
                return original.apply(this, args);
            };
        };
    }
    // A request has a method and request id. We check both primarily to differentiate from
    // a notification which only has method. If in the future we find we should propagate context
    // on notifications too, we can loosen the check.
    static _isJSONRPCRequest(message) {
        return (typeof message === "object" &&
            !!message &&
            "method" in message &&
            typeof message.method === "string" &&
            "id" in message);
    }
}
exports.MCPInstrumentation = MCPInstrumentation;
