// @arizeai/openinference-instrumentation-bedrock/dist/src/attributes/attribute-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setSpanAttribute = setSpanAttribute;
exports.getSystemFromModelId = getSystemFromModelId;
exports.setBasicSpanAttributes = setBasicSpanAttributes;
exports.aggregateSystemPrompts = aggregateSystemPrompts;
exports.aggregateMessages = aggregateMessages;
exports.getAttributesFromMessageContent = getAttributesFromMessageContent;
exports.getAttributesFromMessage = getAttributesFromMessage;
exports.processMessages = processMessages;
exports.extractModelName = extractModelName;
const api_1 = require("@opentelemetry/api");
const openinference_core_1 = require("@arizeai/openinference-core");
const openinference_semantic_conventions_1 = require("@arizeai/openinference-semantic-conventions");
const bedrock_types_1 = require("../types/bedrock-types");
const invoke_model_helpers_1 = require("./invoke-model-helpers");
/**
 * Sets a span attribute only if the value is not null, undefined, or empty string
 * Provides null-safe attribute setting for OpenTelemetry spans to avoid polluting traces with empty values
 *
 * @param span The OpenTelemetry span to set the attribute on
 * @param key The attribute key following OpenInference semantic conventions
 * @param value The attribute value to set, will be skipped if null/undefined/empty
 */
function setSpanAttribute(span, key, value) {
    if (value != null && value !== "") {
        span.setAttribute(key, value);
    }
}
/**
 * Extracts vendor-specific system name from Bedrock model ID
 * Maps model IDs to their corresponding AI system providers
 *
 * @param modelId The full Bedrock model identifier (e.g., "anthropic.claude-3-sonnet-20240229-v1:0")
 * @returns {string} The system provider name (e.g., "anthropic", "meta", "mistral") or "bedrock" as fallback
 */
function getSystemFromModelId(modelId) {
    if (modelId.includes("anthropic"))
        return openinference_semantic_conventions_1.LLMSystem.ANTHROPIC;
    if (modelId.includes("ai21"))
        return openinference_semantic_conventions_1.LLMSystem.AI21;
    if (modelId.includes("amazon"))
        return openinference_semantic_conventions_1.LLMSystem.AMAZON;
    if (modelId.includes("cohere"))
        return openinference_semantic_conventions_1.LLMSystem.COHERE;
    if (modelId.includes("meta"))
        return openinference_semantic_conventions_1.LLMSystem.META;
    if (modelId.includes("mistral"))
        return openinference_semantic_conventions_1.LLMSystem.MISTRALAI;
    return openinference_semantic_conventions_1.LLMSystem.AMAZON;
}
function setBasicSpanAttributes(span, llm_system) {
    setSpanAttribute(span, openinference_semantic_conventions_1.SemanticConventions.LLM_PROVIDER, openinference_semantic_conventions_1.LLMProvider.AWS);
    setSpanAttribute(span, openinference_semantic_conventions_1.SemanticConventions.OPENINFERENCE_SPAN_KIND, openinference_semantic_conventions_1.OpenInferenceSpanKind.LLM);
    setSpanAttribute(span, openinference_semantic_conventions_1.SemanticConventions.LLM_SYSTEM, llm_system);
}
/**
 * Aggregates multiple system prompts into a single string
 * Concatenates all text content from system prompts with double newline separation
 *
 * @param systemPrompts Array of system content blocks from Bedrock Converse API
 * @returns {string} Combined system prompt text with proper formatting
 */
function aggregateSystemPrompts(systemPrompts) {
    return systemPrompts
        .map((prompt) => prompt.text || "")
        .filter(Boolean)
        .join("\n\n");
}
/**
 * Aggregates system prompts with messages into a unified message array
 * System prompts are converted to a single system message at the beginning of the conversation
 *
 * @param systemPrompts Array of system content blocks to convert to system message
 * @param messages Array of conversation messages
 * @returns {Message[]} Combined message array with system prompt as first message if present
 */
function aggregateMessages(systemPrompts = [], messages = []) {
    const aggregated = [];
    if (systemPrompts.length > 0) {
        aggregated.push({
            role: "system",
            content: [{ text: aggregateSystemPrompts(systemPrompts) }],
        });
    }
    return [...aggregated, ...messages];
}
/**
 * Safely converts supported byte-like inputs to base64.
 * Handles: Uint8Array, Buffer.
 * On unsupported inputs, logs a warning and returns undefined.
 */
const toBase64ImageBytes = (0, openinference_core_1.withSafety)({
    fn: (bytes) => {
        if (Buffer.isBuffer(bytes)) {
            return bytes.toString("base64");
        }
        if (bytes instanceof Uint8Array) {
            return Buffer.from(bytes).toString("base64");
        }
        api_1.diag.warn("Unsupported image bytes type encountered");
        return undefined;
    },
    onError: (error) => {
        api_1.diag.warn("Failed to convert image bytes to base64", error);
    },
});
/**
 * Extracts OpenInference semantic convention attributes from message content blocks
 * Handles text, image, and tool content types with appropriate attribute mapping
 *
 * Note: Uses custom content type guards since AWS SDK ContentBlock union types
 * require additional processing for reliable type detection
 *
 * @param content The content block to extract attributes from
 * @returns {Record<string, AttributeValue>} Object containing content-specific attributes
 */
function getAttributesFromMessageContent(content) {
    const attributes = {};
    if ((0, bedrock_types_1.isConverseTextContent)(content)) {
        attributes[openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_TYPE] = "text";
        attributes[openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_TEXT] = content.text;
    }
    else if ((0, bedrock_types_1.isConverseImageContent)(content)) {
        attributes[openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_TYPE] = "image";
        const base64 = toBase64ImageBytes(content.image.source.bytes);
        if (base64) {
            const mimeType = `image/${content.image.format}`;
            attributes[`${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_IMAGE}.${openinference_semantic_conventions_1.SemanticConventions.IMAGE_URL}`] =
                (0, invoke_model_helpers_1.formatImageUrl)({
                    type: "base64",
                    data: base64,
                    media_type: mimeType,
                });
        }
        // Add format attribute for image content
        attributes[`${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_IMAGE}.format`] = content.image.format;
    }
    return attributes;
}
/**
 * Extracts OpenInference semantic convention attributes from a single Bedrock message
 * Handles role, content, tool calls, and tool results following the OpenInference specification
 *
 * @param message The Bedrock message to extract attributes from
 * @returns {Record<string, AttributeValue>} Object containing semantic convention attributes
 */
function getAttributesFromMessage(message) {
    const attributes = {};
    if (message.role) {
        attributes[openinference_semantic_conventions_1.SemanticConventions.MESSAGE_ROLE] = message.role;
    }
    if (message.content) {
        // Check if this is a simple single text content case
        if (message.content.length === 1 && (0, bedrock_types_1.isConverseTextContent)(message.content[0])) {
            // Use simple format for single text content
            attributes[openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT] = message.content[0].text;
        }
        else {
            // Use complex format for multi-modal or multi-content messages
            let toolCallIndex = 0;
            for (const [index, content] of message.content.entries()) {
                // Process content as our custom types for attribute extraction
                const contentAttributes = getAttributesFromMessageContent(content);
                for (const key in contentAttributes) {
                    attributes[`${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENTS}.${index}.${key}`] =
                        contentAttributes[key];
                }
                // Handle tool calls at the message level using proper semantic conventions
                if ((0, bedrock_types_1.isConverseToolUseContent)(content)) {
                    const toolCallPrefix = `${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALLS}.${toolCallIndex}`;
                    attributes[`${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_ID}`] =
                        content.toolUse.toolUseId;
                    attributes[`${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_NAME}`] =
                        content.toolUse.name;
                    const argsJson = (0, openinference_core_1.safelyJSONStringify)(content.toolUse.input);
                    if (argsJson) {
                        attributes[`${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_ARGUMENTS_JSON}`] = argsJson;
                    }
                    toolCallIndex++;
                }
                else if ((0, bedrock_types_1.isConverseToolResultContent)(content)) {
                    attributes[openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALL_ID] = content.toolResult.toolUseId;
                }
            }
        }
    }
    return attributes;
}
/**
 * Processes multiple messages and sets OpenInference attributes on span with proper indexing
 * Iterates through messages array and applies semantic convention attributes with message index
 *
 * @param params Object containing processing parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.messages Array of messages to process
 * @param params.baseKey Base semantic convention key (either input or output messages)
 */
function processMessages({ span, messages, baseKey, }) {
    for (const [index, message] of messages.entries()) {
        const messageAttributes = getAttributesFromMessage(message);
        for (const [key, value] of Object.entries(messageAttributes)) {
            setSpanAttribute(span, `${baseKey}.${index}.${key}`, value);
        }
    }
}
/**
 * Extracts clean model name from full Bedrock model ID
 * Removes vendor prefix and version suffixes to get the base model name
 *
 * @param modelId The full Bedrock model identifier
 * @returns {string} The cleaned model name (e.g., "claude-3-sonnet" from "anthropic.claude-3-sonnet-20240229-v1:0")
 */
function extractModelName(modelId) {
    const parts = modelId.split(".");
    if (parts.length > 1) {
        const modelPart = parts[1];
        if (modelId.includes("anthropic")) {
            const versionIndex = modelPart.indexOf("-v");
            if (versionIndex > 0) {
                return modelPart.substring(0, versionIndex);
            }
        }
        return modelPart;
    }
    return modelId;
}
