// @arizeai/openinference-instrumentation-bedrock/dist/src/attributes/converse-streaming-response-attributes.js
"use strict";
/**
 * Streaming response attribute extraction for AWS Bedrock Converse API instrumentation
 *
 * Handles extraction of semantic convention attributes from streaming Converse responses including:
 * - Streaming content processing for Converse events
 * - Tool call extraction from streams
 * - Token usage accumulation from metadata events
 * - Real-time span attribute setting
 * - Safe stream splitting with original stream preservation
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.consumeConverseStreamChunks = void 0;
const api_1 = require("@opentelemetry/api");
const openinference_core_1 = require("@arizeai/openinference-core");
const openinference_semantic_conventions_1 = require("@arizeai/openinference-semantic-conventions");
const bedrock_types_1 = require("../types/bedrock-types");
const attribute_helpers_1 = require("./attribute-helpers");
/**
 * Resolves the target tool use id from either an explicit id or a content block index.
 *
 * @param params.id Optional explicit tool id carried on the event
 * @param params.contentBlockIndex Optional content block index to map into an id
 * @param params.indexMap Optional index->id mapping established at tool start
 * @returns The resolved tool id if available
 */
function resolveToolUseId({ id, contentBlockIndex, indexMap, }) {
    if (id)
        return id;
    if (contentBlockIndex !== undefined && indexMap) {
        return indexMap[contentBlockIndex];
    }
}
/**
 * Starts tracking a tool call encountered in a Converse stream.
 *
 * Records the mapping between a content block index and a toolUseId so that
 * subsequent input chunks can be correlated, and initializes a new tool call
 * entry with an empty input buffer for incremental JSON accumulation.
 *
 * @param state Stream processing state to update
 * @param params.id The tool use identifier from the stream event
 * @param params.name The tool name provided by the model
 * @param params.contentBlockIndex Optional index to correlate future input chunks
 */
function startToolCall(state, { id, name, contentBlockIndex }) {
    var _a;
    if (contentBlockIndex !== undefined) {
        (_a = state.toolUseIdByIndex) !== null && _a !== void 0 ? _a : (state.toolUseIdByIndex = {});
        state.toolUseIdByIndex[contentBlockIndex] = id;
    }
    state.toolCalls.push({ id, name, input: {}, partialJsonInput: "" });
}
/**
 * Appends a partial JSON input chunk for the current tool call.
 *
 * Resolves the target tool call either by explicit id or by the content block
 * index mapping established at tool start. Accumulates the chunk into a
 * string buffer and attempts to parse it into a JSON object; on parse errors,
 * accumulation continues until valid JSON is formed by subsequent chunks.
 *
 * @param state Stream processing state to update
 * @param params.chunk Partial JSON string for tool input
 * @param params.contentBlockIndex Optional index to resolve the tool id
 * @param params.id Optional explicit tool id
 */
function appendToolInputChunk(state, { chunk, contentBlockIndex, id }) {
    var _a;
    const targetId = resolveToolUseId({
        id,
        contentBlockIndex,
        indexMap: state.toolUseIdByIndex,
    });
    if (!targetId)
        return;
    const tool = state.toolCalls.find((t) => t.id === targetId);
    if (!tool)
        return;
    tool.partialJsonInput = ((_a = tool.partialJsonInput) !== null && _a !== void 0 ? _a : "") + chunk;
    const parsed = (0, openinference_core_1.safelyJSONParse)(tool.partialJsonInput);
    if (parsed != null && (0, openinference_core_1.isObjectWithStringKeys)(parsed)) {
        tool.input = parsed;
    }
}
/**
 * Processes Converse stream chunks and updates processing state
 * Handles messageStart, contentBlockDelta, toolUse events, and metadata with proper accumulation
 *
 * @param data The parsed Converse stream event data
 * @param state Current processing state to update
 * @returns Updated processing state
 */
function processConverseStreamChunk(data, state) {
    var _a, _b, _c;
    const ev = (0, bedrock_types_1.toNormalizedConverseStreamEvent)(data);
    if (!ev)
        return;
    switch (ev.kind) {
        case "messageStart":
            return;
        case "messageStop":
            state.stopReason = ev.stopReason;
            return;
        case "textDelta":
            state.outputText += ev.text;
            return;
        case "toolUseStart":
            startToolCall(state, ev);
            return;
        case "toolUseInputChunk":
            appendToolInputChunk(state, ev);
            return;
        case "metadata":
            state.usage = {
                inputTokens: (_a = ev.usage.inputTokens) !== null && _a !== void 0 ? _a : state.usage.inputTokens,
                outputTokens: (_b = ev.usage.outputTokens) !== null && _b !== void 0 ? _b : state.usage.outputTokens,
                totalTokens: (_c = ev.usage.totalTokens) !== null && _c !== void 0 ? _c : state.usage.totalTokens,
            };
            return;
        default:
            (0, openinference_core_1.assertUnreachable)(ev);
    }
}
/**
 * Sets streaming output attributes on the OpenTelemetry span for Converse streams
 * Processes accumulated content and usage data into semantic convention attributes
 *
 * @param params Object containing span and accumulated streaming data
 */
function setConverseStreamingOutputAttributes({ span, outputText, toolCalls, usage, stopReason, }) {
    // Create the output value structure similar to converse response format
    // Convert usage from camelCase to snake_case for consistency
    const normalizedUsage = Object.assign(Object.assign(Object.assign({}, (usage.inputTokens !== undefined && { input_tokens: usage.inputTokens })), (usage.outputTokens !== undefined && {
        output_tokens: usage.outputTokens,
    })), (usage.totalTokens !== undefined && { total_tokens: usage.totalTokens }));
    // Clean up tool calls - drop partialJsonInput from final output
    const cleanedToolCalls = toolCalls.map(({ id, name, input }) => ({
        id,
        name,
        input,
    }));
    const outputValue = Object.assign({ text: outputText || "", tool_calls: cleanedToolCalls, usage: normalizedUsage, streaming: true }, (stopReason && { stop_reason: stopReason }));
    // Set output value as JSON (matching converse response behavior)
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.OUTPUT_VALUE, (0, openinference_core_1.safelyJSONStringify)(outputValue));
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.OUTPUT_MIME_TYPE, openinference_semantic_conventions_1.MimeType.JSON);
    // Set the message role (always assistant for converse responses)
    (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_OUTPUT_MESSAGES}.0.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_ROLE}`, "assistant");
    // Set the main accumulated text content
    if (outputText) {
        (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_OUTPUT_MESSAGES}.0.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT}`, outputText);
    }
    // Set tool call attributes with sequential indexing
    toolCalls.forEach((toolCall, toolCallIndex) => {
        const toolCallPrefix = `${openinference_semantic_conventions_1.SemanticConventions.LLM_OUTPUT_MESSAGES}.0.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALLS}.${toolCallIndex}`;
        (0, attribute_helpers_1.setSpanAttribute)(span, `${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_ID}`, toolCall.id);
        (0, attribute_helpers_1.setSpanAttribute)(span, `${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_NAME}`, toolCall.name);
        (0, attribute_helpers_1.setSpanAttribute)(span, `${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_ARGUMENTS_JSON}`, (0, openinference_core_1.safelyJSONStringify)(toolCall.input));
    });
    // Set stop reason attribute if available
    if (stopReason) {
        (0, attribute_helpers_1.setSpanAttribute)(span, "llm.stop_reason", stopReason);
    }
    // Set usage attributes
    if (usage.inputTokens !== undefined) {
        (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_TOKEN_COUNT_PROMPT, usage.inputTokens);
    }
    if (usage.outputTokens !== undefined) {
        (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_TOKEN_COUNT_COMPLETION, usage.outputTokens);
    }
    if (usage.totalTokens !== undefined) {
        (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_TOKEN_COUNT_TOTAL, usage.totalTokens);
    }
}
/**
 * Consumes AWS Bedrock Converse streaming response chunks and extracts attributes for OpenTelemetry span.
 * Supports the unified Converse API streaming format with messageStart, contentBlockDelta, metadata events.
 *
 * This function processes the Converse streaming format and accumulates response content,
 * setting appropriate semantic convention attributes on the provided span. It runs in the
 * background without blocking the user's stream consumption.
 *
 * @param stream - The Converse response stream (AsyncIterable), typically from safelySplitStream()
 * @param span - The OpenTelemetry span to set attributes on
 * @throws {Error} If critical stream processing errors occur
 *
 * @example
 * ```typescript
 * // Background processing after stream splitting
 * const { instrumentationStream, userStream } = safelySplitStream(response.stream);
 * consumeConverseStreamChunks({ stream: instrumentationStream, span })
 *   .then(() => span.end())
 *   .catch(error => { span.recordException(error); span.end(); });
 * ```
 */
exports.consumeConverseStreamChunks = (0, openinference_core_1.withSafety)({
    fn: (_a) => __awaiter(void 0, [_a], void 0, function* ({ stream, span }) {
        var _b, stream_1, stream_1_1;
        var _c, e_1, _d, _e;
        const state = {
            outputText: "",
            toolCalls: [],
            usage: {},
        };
        try {
            for (_b = true, stream_1 = __asyncValues(stream); stream_1_1 = yield stream_1.next(), _c = stream_1_1.done, !_c; _b = true) {
                _e = stream_1_1.value;
                _b = false;
                const chunk = _e;
                // ConverseStream returns events directly as objects, no need to decode bytes
                if ((0, bedrock_types_1.isValidConverseStreamEventData)(chunk)) {
                    processConverseStreamChunk(chunk, state);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (!_b && !_c && (_d = stream_1.return)) yield _d.call(stream_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        setConverseStreamingOutputAttributes({
            span,
            outputText: state.outputText,
            toolCalls: state.toolCalls,
            usage: state.usage,
            stopReason: state.stopReason,
        });
    }),
    onError: (error) => {
        api_1.diag.warn("Error consuming Converse stream chunks:", error);
        throw error;
    },
});
