// @arizeai/openinference-instrumentation-bedrock/dist/src/attributes/converse-response-attributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractConverseResponseAttributes = void 0;
const api_1 = require("@opentelemetry/api");
const openinference_core_1 = require("@arizeai/openinference-core");
const openinference_semantic_conventions_1 = require("@arizeai/openinference-semantic-conventions");
const bedrock_types_1 = require("../types/bedrock-types");
const attribute_helpers_1 = require("./attribute-helpers");
/**
 * Type guard to safely validate ConverseResponse structure
 * Ensures the response object has the expected output structure for processing
 *
 * @param response The response object to validate
 * @returns {boolean} True if response is a valid ConverseResponse, false otherwise
 */
function isConverseResponse(response) {
    if (!(0, openinference_core_1.isObjectWithStringKeys)(response)) {
        return false;
    }
    return "output" in response && typeof response.output === "object" && response.output !== null;
}
/**
 * Extracts base response attributes for OpenInference semantic conventions
 * Sets output value, MIME type, and stop reason attributes
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.response The ConverseResponse to extract attributes from
 */
function extractBaseResponseAttributes({ span, response, }) {
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.OUTPUT_VALUE, JSON.stringify(response));
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.OUTPUT_MIME_TYPE, openinference_semantic_conventions_1.MimeType.JSON);
    (0, attribute_helpers_1.setSpanAttribute)(span, "llm.stop_reason", response.stopReason);
}
/**
 * Extracts output message attributes from Converse response
 * Processes the response message and sets OpenInference message attributes
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.response The ConverseResponse containing output message
 */
function extractOutputMessagesAttributes({ span, response, }) {
    const output = response.output;
    if (!output || !("message" in output))
        return;
    const outputMessage = output.message;
    if (!outputMessage)
        return;
    (0, attribute_helpers_1.processMessages)({
        span,
        messages: [outputMessage],
        baseKey: openinference_semantic_conventions_1.SemanticConventions.LLM_OUTPUT_MESSAGES,
    });
}
/**
 * Extracts tool call attributes from response content blocks
 * Identifies and processes tool use blocks in the response message
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.response The ConverseResponse containing tool calls
 */
function extractToolCallAttributes({ span, response, }) {
    const output = response.output;
    if (!output || !("message" in output))
        return;
    const outputMessage = output.message;
    if (!outputMessage || !outputMessage.content)
        return;
    const toolUseBlocks = outputMessage.content.filter(bedrock_types_1.isConverseToolUseContent);
    toolUseBlocks.forEach((content, toolCallIndex) => {
        const toolUse = content.toolUse;
        if (toolUse) {
            (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_OUTPUT_MESSAGES}.0.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALLS}.${toolCallIndex}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_NAME}`, toolUse.name);
            (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_OUTPUT_MESSAGES}.0.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALLS}.${toolCallIndex}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_ARGUMENTS_JSON}`, JSON.stringify(toolUse.input));
            (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_OUTPUT_MESSAGES}.0.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALLS}.${toolCallIndex}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_ID}`, toolUse.toolUseId);
        }
    });
}
/**
 * Extracts token usage statistics with null-safe handling for missing token counts
 * Sets input, output, and total token count attributes following OpenInference conventions
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.response The ConverseResponse containing usage statistics
 */
function extractUsageAttributes({ span, response, }) {
    const usage = response.usage;
    if (!usage)
        return;
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_TOKEN_COUNT_PROMPT, usage.inputTokens);
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_TOKEN_COUNT_COMPLETION, usage.outputTokens);
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_TOKEN_COUNT_TOTAL, usage.totalTokens);
}
/**
 * Extracts semantic convention attributes from Converse response and sets them on the span
 * Main entry point for processing Converse API responses with comprehensive error handling
 *
 * Processes:
 * - Base response attributes (output value, MIME type, stop reason)
 * - Output messages with content structure
 * - Tool call attributes from response blocks
 * - Token usage statistics
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.response The ConverseResponse to extract attributes from
 */
exports.extractConverseResponseAttributes = (0, openinference_core_1.withSafety)({
    fn: ({ span, response }) => {
        if (!isConverseResponse(response)) {
            return;
        }
        extractBaseResponseAttributes({ span, response });
        extractOutputMessagesAttributes({ span, response });
        extractToolCallAttributes({ span, response });
        extractUsageAttributes({ span, response });
    },
    onError: (error) => {
        api_1.diag.warn("Error extracting Converse response attributes:", error);
    },
});
