// @arizeai/openinference-instrumentation-bedrock/dist/src/attributes/converse-request-attributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractConverseRequestAttributes = void 0;
const api_1 = require("@opentelemetry/api");
const openinference_core_1 = require("@arizeai/openinference-core");
const openinference_semantic_conventions_1 = require("@arizeai/openinference-semantic-conventions");
const attribute_helpers_1 = require("./attribute-helpers");
/**
 * Type guard to safely validate ConverseRequest structure
 * Ensures the input object has required modelId and messages properties for processing
 *
 * @param input The input object to validate
 * @returns {boolean} True if input is a valid ConverseRequest, false otherwise
 */
function isConverseRequest(input) {
    if (!(0, openinference_core_1.isObjectWithStringKeys)(input)) {
        return false;
    }
    return ("modelId" in input &&
        typeof input.modelId === "string" &&
        "messages" in input &&
        Array.isArray(input.messages));
}
/**
 * Extracts base request attributes for OpenInference semantic conventions
 * Sets fundamental LLM attributes including model, system, provider, and invocation parameters
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.input The validated ConverseRequest to extract attributes from
 */
function extractBaseRequestAttributes({ span, input, }) {
    const modelId = input.modelId || "unknown";
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_MODEL_NAME, (0, attribute_helpers_1.extractModelName)(modelId));
    if (input.inferenceConfig && Object.keys(input.inferenceConfig).length > 0) {
        (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_INVOCATION_PARAMETERS, JSON.stringify(input.inferenceConfig));
    }
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.INPUT_VALUE, JSON.stringify(input));
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.INPUT_MIME_TYPE, openinference_semantic_conventions_1.MimeType.JSON);
}
/**
 * Extracts input messages attributes with system prompt aggregation
 * Processes system prompts and conversation messages into OpenInference message format
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.input The ConverseRequest containing messages and system prompts
 */
function extractInputMessagesAttributes({ span, input, }) {
    const systemPrompts = input.system || [];
    const messages = input.messages || [];
    const aggregatedMessages = (0, attribute_helpers_1.aggregateMessages)(systemPrompts, messages);
    (0, attribute_helpers_1.processMessages)({
        span,
        messages: aggregatedMessages,
        baseKey: openinference_semantic_conventions_1.SemanticConventions.LLM_INPUT_MESSAGES,
    });
}
/**
 * Extracts tool configuration attributes from Converse request
 * Processes tool definitions and sets them as JSON schema attributes
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.input The ConverseRequest containing tool configuration
 */
function extractInputToolAttributes({ span, input }) {
    const toolConfig = input.toolConfig;
    if (!(toolConfig === null || toolConfig === void 0 ? void 0 : toolConfig.tools))
        return;
    toolConfig.tools.forEach((tool, index) => {
        (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_TOOLS}.${index}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_JSON_SCHEMA}`, JSON.stringify(tool));
    });
}
/**
 * Extracts semantic convention attributes from Converse command and sets them on the span
 * Main entry point for processing Converse API requests with comprehensive error handling
 *
 * Processes:
 * - Base model and system attributes
 * - Input messages with system prompt aggregation
 * - Tool configuration definitions
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.command The ConverseCommand to extract attributes from
 */
exports.extractConverseRequestAttributes = (0, openinference_core_1.withSafety)({
    fn: ({ span, command, }) => {
        const input = command.input;
        if (!input || !isConverseRequest(input)) {
            return;
        }
        extractBaseRequestAttributes({ span, input });
        extractInputMessagesAttributes({ span, input });
        extractInputToolAttributes({ span, input });
    },
    onError: (error) => {
        api_1.diag.warn("Error extracting Converse request attributes:", error);
    },
});
