// @arizeai/openinference-instrumentation-bedrock/dist/src/attributes/invoke-model-request-attributes.js
"use strict";
/**
 * Request attribute extraction for AWS Bedrock instrumentation
 *
 * Handles extraction of semantic convention attributes from InvokeModel requests including:
 * - Base model and system attributes
 * - Input message processing
 * - Tool definition processing
 * - Invocation parameters
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractInvokeModelRequestAttributes = void 0;
const api_1 = require("@opentelemetry/api");
const openinference_core_1 = require("@arizeai/openinference-core");
const openinference_semantic_conventions_1 = require("@arizeai/openinference-semantic-conventions");
const bedrock_types_1 = require("../types/bedrock-types");
const attribute_helpers_1 = require("./attribute-helpers");
const invoke_model_helpers_1 = require("./invoke-model-helpers");
// Helper functions
/**
 * Adds message attributes to the span following OpenInference conventions
 * Processes role, content, tool calls, and tool results with proper indexing
 *
 * @param params Object containing message processing parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.message The Bedrock message to process
 * @param params.index The message index in the conversation
 */
function addMessageAttributes({ span, message, index, }) {
    if (!message.role)
        return;
    (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_INPUT_MESSAGES}.${index}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_ROLE}`, message.role);
    if (typeof message.content === "string") {
        (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_INPUT_MESSAGES}.${index}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT}`, message.content);
    }
    else {
        addMessageContentAttributes({ span, message, messageIndex: index });
    }
    // Handle tool calls and results at the message level
    if (Array.isArray(message.content)) {
        handleToolCallsInMessage({ span, message, messageIndex: index });
        handleToolResultsInMessage({ span, message, messageIndex: index });
    }
}
/**
 * Handles tool calls within a message using OpenInference semantic conventions
 * Extracts tool call information and sets appropriate attributes with indexing
 *
 * @param params Object containing tool call processing parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.message The Bedrock message containing tool calls
 * @param params.messageIndex The message index in the conversation
 */
function handleToolCallsInMessage({ span, message, messageIndex, }) {
    if (!Array.isArray(message.content))
        return;
    let toolCallIndex = 0;
    message.content.forEach((content) => {
        if ((0, bedrock_types_1.isToolUseContent)(content)) {
            const toolCallPrefix = `${openinference_semantic_conventions_1.SemanticConventions.LLM_INPUT_MESSAGES}.${messageIndex}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALLS}.${toolCallIndex}`;
            (0, attribute_helpers_1.setSpanAttribute)(span, `${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_ID}`, content.id);
            (0, attribute_helpers_1.setSpanAttribute)(span, `${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_NAME}`, content.name);
            (0, attribute_helpers_1.setSpanAttribute)(span, `${toolCallPrefix}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_CALL_FUNCTION_ARGUMENTS_JSON}`, JSON.stringify(content.input));
            toolCallIndex++;
        }
    });
}
/**
 * Handles tool results within a message using OpenInference semantic conventions
 * Extracts tool result information and sets tool call ID references
 *
 * @param params Object containing tool result processing parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.message The Bedrock message containing tool results
 * @param params.messageIndex The message index in the conversation
 */
function handleToolResultsInMessage({ span, message, messageIndex, }) {
    if (!Array.isArray(message.content))
        return;
    const toolResultBlocks = (0, invoke_model_helpers_1.extractToolResultBlocks)(message.content);
    toolResultBlocks.forEach((contentBlock) => {
        (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_INPUT_MESSAGES}.${messageIndex}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_TOOL_CALL_ID}`, contentBlock.tool_use_id);
    });
}
/**
 * Adds detailed message content structure attributes for multi-modal content
 * Processes text, image, and other content types with appropriate OpenInference attributes
 *
 * @param params Object containing content processing parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.message The Bedrock message containing structured content
 * @param params.messageIndex The message index in the conversation
 */
function addMessageContentAttributes({ span, message, messageIndex, }) {
    if (!message.content)
        return;
    if (Array.isArray(message.content)) {
        message.content.forEach((content, contentIndex) => {
            const contentPrefix = `${openinference_semantic_conventions_1.SemanticConventions.LLM_INPUT_MESSAGES}.${messageIndex}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENTS}.${contentIndex}`;
            if ((0, bedrock_types_1.isTextContent)(content)) {
                (0, attribute_helpers_1.setSpanAttribute)(span, `${contentPrefix}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_TYPE}`, "text");
                (0, attribute_helpers_1.setSpanAttribute)(span, `${contentPrefix}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_TEXT}`, content.text);
            }
            else if ((0, bedrock_types_1.isImageContent)(content)) {
                (0, attribute_helpers_1.setSpanAttribute)(span, `${contentPrefix}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_TYPE}`, "image");
                const imageUrl = (0, invoke_model_helpers_1.formatImageUrl)(content.source);
                (0, attribute_helpers_1.setSpanAttribute)(span, `${contentPrefix}.${openinference_semantic_conventions_1.SemanticConventions.MESSAGE_CONTENT_IMAGE}.${openinference_semantic_conventions_1.SemanticConventions.IMAGE_URL}`, imageUrl);
            }
        });
    }
}
/**
 * Extracts base request attributes for OpenInference semantic conventions
 * Sets fundamental LLM attributes including model, system, provider, and invocation parameters
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.command The InvokeModelCommand containing model ID
 * @param params.requestBody The parsed request body containing model parameters
 */
function extractBaseRequestAttributes({ span, command, requestBody, system, }) {
    var _a;
    const modelId = ((_a = command.input) === null || _a === void 0 ? void 0 : _a.modelId) || "unknown";
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_MODEL_NAME, (0, attribute_helpers_1.extractModelName)(modelId));
    const inputValue = JSON.stringify(requestBody);
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.INPUT_VALUE, inputValue);
    (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.INPUT_MIME_TYPE, openinference_semantic_conventions_1.MimeType.JSON);
    const invocationParams = (0, invoke_model_helpers_1.extractInvocationParameters)(requestBody, system);
    if (Object.keys(invocationParams).length > 0) {
        (0, attribute_helpers_1.setSpanAttribute)(span, openinference_semantic_conventions_1.SemanticConventions.LLM_INVOCATION_PARAMETERS, JSON.stringify(invocationParams));
    }
}
/**
 * Extracts input messages attributes from InvokeModel request body
 * Processes message array and sets OpenInference input message attributes with proper indexing
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.requestBody The parsed request body containing messages array
 */
function extractInputMessagesAttributes({ span, requestBody, system, }) {
    const messages = (0, invoke_model_helpers_1.normalizeRequestContentBlocks)(requestBody, system);
    if (messages && Array.isArray(messages)) {
        messages.forEach((message, index) => {
            addMessageAttributes({ span, message, index });
        });
    }
}
/**
 * Extracts input tool attributes from InvokeModel request body
 * Processes tool definitions array and sets them as JSON schema attributes
 * Out of supported models, Anthropic, Amazon Nova, and Mistral Chat Completion
 * support tool calls.
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.requestBody The parsed request body containing tools array
 */
function extractInputToolAttributes({ span, requestBody, system, }) {
    if (system === openinference_semantic_conventions_1.LLMSystem.AMAZON &&
        requestBody.toolConfig &&
        (0, openinference_core_1.isObjectWithStringKeys)(requestBody.toolConfig) &&
        requestBody.toolConfig.tools &&
        Array.isArray(requestBody.toolConfig.tools)) {
        requestBody.toolConfig.tools.forEach((tool, index) => {
            // Record each tool element verbatim (keeping the Converse tagged-union
            // envelope) to match the OpenAI/Anthropic instrumentors and the Converse
            // extractor, and to preserve non-toolSpec members such as systemTool.
            if ((0, openinference_core_1.isObjectWithStringKeys)(tool)) {
                (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_TOOLS}.${index}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_JSON_SCHEMA}`, JSON.stringify(tool));
            }
        });
    }
    else if (requestBody.tools && Array.isArray(requestBody.tools)) {
        requestBody.tools.forEach((tool, index) => {
            // Both Anthropic and Mistral Chat Completion have the tool definitions in the same place
            (0, attribute_helpers_1.setSpanAttribute)(span, `${openinference_semantic_conventions_1.SemanticConventions.LLM_TOOLS}.${index}.${openinference_semantic_conventions_1.SemanticConventions.TOOL_JSON_SCHEMA}`, JSON.stringify(tool));
        });
    }
}
/**
 * Extracts semantic convention attributes from InvokeModel request command
 * Main entry point for processing InvokeModel API requests with comprehensive error handling
 *
 * Processes:
 * - Request body parsing from multiple formats
 * - Base model and system attributes
 * - Input messages with multi-modal content
 * - Tool definitions
 *
 * @param params Object containing extraction parameters
 * @param params.span The OpenTelemetry span to set attributes on
 * @param params.command The InvokeModelCommand to extract attributes from
 */
exports.extractInvokeModelRequestAttributes = (0, openinference_core_1.withSafety)({
    fn: ({ span, command, system, }) => {
        const requestBody = (0, invoke_model_helpers_1.parseRequestBody)(command);
        if (!requestBody) {
            return;
        }
        extractBaseRequestAttributes({ span, command, requestBody, system });
        extractInputMessagesAttributes({ span, requestBody, system });
        extractInputToolAttributes({ span, requestBody, system });
    },
    onError: (error) => {
        api_1.diag.warn("Error extracting InvokeModel request attributes:", error);
    },
});
