// @arizeai/openinference-core/dist/esm/trace/trace-config/OITracer.js
import { context as apiContext } from "@opentelemetry/api";
import { getAttributesFromContext } from "../contextAttributes.js";
import { OISpan } from "./OISpan.js";
import { generateTraceConfig } from "./traceConfig.js";
/**
 * Formats the params for the startActiveSpan method
 * The method has multiple overloads, so we need to format the arguments
 * Taken from @see https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/Tracer.ts#L220C3-L235C6
 *
 */
function formatStartActiveSpanParams(arg2, arg3, arg4) {
    let opts;
    let ctx;
    let fn;
    if (typeof arg2 === "function") {
        fn = arg2;
    }
    else if (typeof arg3 === "function") {
        opts = arg2;
        fn = arg3;
    }
    else {
        opts = arg2;
        ctx = arg3;
        fn = arg4;
    }
    opts = opts ?? {};
    ctx = ctx ?? apiContext.active();
    return { opts, ctx, fn };
}
/**
 * A wrapper around the OpenTelemetry {@link Tracer} interface that masks sensitive information based on the passed in {@link TraceConfig}.
 */
export class OITracer {
    /**
     * @param params - The parameters for the OITracer constructor
     * @param params.tracer - The OpenTelemetry {@link Tracer} to wrap
     * @param params.traceConfig - The {@link TraceConfigOptions} to set to control the behavior of the tracer
     */
    constructor({ tracer, traceConfig }) {
        this.tracer = tracer;
        this.config = generateTraceConfig(traceConfig);
    }
    startActiveSpan(name, arg2, arg3, arg4) {
        const formattedArgs = formatStartActiveSpanParams(arg2, arg3, arg4);
        if (formattedArgs == null) {
            return;
        }
        const { opts, ctx, fn } = formattedArgs;
        const { attributes } = opts ?? {};
        const contextAttributes = getAttributesFromContext(ctx);
        const mergedAttributes = { ...contextAttributes, ...attributes };
        return this.tracer.startActiveSpan(name, { ...opts, attributes: undefined }, ctx, (span) => {
            const openInferenceSpan = new OISpan({
                span,
                config: this.config,
            });
            openInferenceSpan.setAttributes(mergedAttributes);
            return fn(openInferenceSpan);
        });
    }
    startSpan(name, options, context) {
        const attributes = options?.attributes;
        const ctx = context ?? apiContext.active();
        const contextAttributes = getAttributesFromContext(ctx);
        const mergedAttributes = { ...contextAttributes, ...attributes };
        const span = new OISpan({
            span: this.tracer.startSpan(name, { ...options, attributes: undefined }, ctx),
            config: this.config,
        });
        span.setAttributes(mergedAttributes);
        return span;
    }
}
