// @arizeai/openinference-instrumentation-bedrock/dist/src/instrumentation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BedrockInstrumentation = void 0;
exports.isPatched = isPatched;
const types_1 = require("util/types");
const client_bedrock_runtime_1 = require("@aws-sdk/client-bedrock-runtime");
const api_1 = require("@opentelemetry/api");
const instrumentation_1 = require("@opentelemetry/instrumentation");
const openinference_core_1 = require("@arizeai/openinference-core");
const attribute_helpers_1 = require("./attributes/attribute-helpers");
const converse_request_attributes_1 = require("./attributes/converse-request-attributes");
const converse_response_attributes_1 = require("./attributes/converse-response-attributes");
const converse_streaming_response_attributes_1 = require("./attributes/converse-streaming-response-attributes");
const invoke_model_request_attributes_1 = require("./attributes/invoke-model-request-attributes");
const invoke_model_response_attributes_1 = require("./attributes/invoke-model-response-attributes");
const invoke_model_streaming_response_attributes_1 = require("./attributes/invoke-model-streaming-response-attributes");
const version_1 = require("./version");
const MODULE_NAME = "@aws-sdk/client-bedrock-runtime";
const INSTRUMENTATION_NAME = "@arizeai/openinference-instrumentation-bedrock";
const INSTRUMENTATION_VERSION = version_1.VERSION;
/**
 * Track if the Bedrock instrumentation is patched
 */
let _isBedrockPatched = false;
/**
 * Type guard that checks if the Bedrock instrumentation is currently enabled
 * @returns {boolean} True if instrumentation is active, false otherwise
 */
function isPatched() {
    return _isBedrockPatched;
}
/**
 * An auto instrumentation class for AWS Bedrock that creates {@link https://github.com/Arize-ai/openinference/blob/main/spec/semantic_conventions.md|OpenInference} Compliant spans for Bedrock API calls
 *
 * Supports instrumentation of:
 * - InvokeModel commands (synchronous)
 * - InvokeModelWithResponseStream commands (streaming)
 * - Converse commands (conversation API)
 * - ConverseStream commands (streaming conversation API)
 *
 * @param instrumentationConfig The config for the instrumentation @see {@link InstrumentationConfig}
 * @param traceConfig The OpenInference trace configuration. Can be used to mask or redact sensitive information on spans. @see {@link TraceConfigOptions}
 */
class BedrockInstrumentation extends instrumentation_1.InstrumentationBase {
    constructor({ instrumentationConfig, traceConfig, tracerProvider, } = {}) {
        var _a, _b;
        super(INSTRUMENTATION_NAME, INSTRUMENTATION_VERSION, Object.assign({}, instrumentationConfig));
        this.tracerProvider = tracerProvider;
        this.traceConfig = traceConfig;
        this.oiTracer = new openinference_core_1.OITracer({
            tracer: (_b = (_a = this.tracerProvider) === null || _a === void 0 ? void 0 : _a.getTracer(INSTRUMENTATION_NAME)) !== null && _b !== void 0 ? _b : this.tracer,
            traceConfig,
        });
    }
    /**
     * Manually instruments the BedrockRuntimeClient, this allows for guaranteed patching of the module when import order is hard to control.
     */
    manuallyInstrument(module) {
        api_1.diag.debug(`Manually instrumenting ${MODULE_NAME}`);
        this.patch(module);
    }
    /**
     * Initializes the instrumentation module definition for AWS SDK Bedrock Runtime
     * @returns {InstrumentationModuleDefinition<BedrockModuleExports>[]} Array containing the module definition
     */
    init() {
        const module = new instrumentation_1.InstrumentationNodeModuleDefinition(MODULE_NAME, ["^3.0.0"], this.patch.bind(this), this.unpatch.bind(this));
        return [module];
    }
    get tracer() {
        if (this.tracerProvider) {
            return this.tracerProvider.getTracer(this.instrumentationName);
        }
        return super.tracer;
    }
    setTracerProvider(tracerProvider) {
        super.setTracerProvider(tracerProvider);
        this.tracerProvider = tracerProvider;
        this.oiTracer = new openinference_core_1.OITracer({
            tracer: this.tracer,
            traceConfig: this.traceConfig,
        });
    }
    /**
     * Patches the BedrockRuntimeClient to intercept and instrument send() method calls
     * Wraps the send method to capture InvokeModel, InvokeModelWithResponseStream, and Converse commands
     *
     * @param moduleExports The module exports from @aws-sdk/client-bedrock-runtime
     * @param moduleVersion The version of the module being patched
     * @returns {BedrockModuleExports} The patched module exports
     */
    patch(moduleExports, moduleVersion) {
        api_1.diag.debug(`Applying patch for ${MODULE_NAME}@${moduleVersion}`);
        if (moduleExports === null || moduleExports === void 0 ? void 0 : moduleExports.BedrockRuntimeClient) {
            // eslint-disable-next-line @typescript-eslint/no-this-alias
            const instrumentation = this;
            // Wrap the client's send method to intercept commands
            this._wrap(moduleExports.BedrockRuntimeClient.prototype, "send", (original) => {
                return function patchedSend(...args) {
                    const command = args[0];
                    if (command instanceof client_bedrock_runtime_1.InvokeModelCommand) {
                        return instrumentation.handleInvokeModelCommand(args, command, original, this);
                    }
                    if (command instanceof client_bedrock_runtime_1.InvokeModelWithResponseStreamCommand) {
                        return instrumentation.handleInvokeModelWithResponseStreamCommand(args, command, original, this);
                    }
                    if (command instanceof client_bedrock_runtime_1.ConverseCommand) {
                        return instrumentation.handleConverseCommand(args, command, original, this);
                    }
                    if (command instanceof client_bedrock_runtime_1.ConverseStreamCommand) {
                        return instrumentation.handleConverseStreamCommand(args, command, original, this);
                    }
                    // Pass through other commands without instrumentation
                    return original.apply(this, args);
                };
            });
            _isBedrockPatched = true;
        }
        return moduleExports;
    }
    /**
     * Removes the instrumentation patch from the BedrockRuntimeClient
     * Unwraps the send method and resets the patched state
     *
     * @param moduleExports The module exports from @aws-sdk/client-bedrock-runtime
     * @param moduleVersion The version of the module being unpatched
     */
    unpatch(moduleExports, moduleVersion) {
        api_1.diag.debug(`Removing patch for ${MODULE_NAME}@${moduleVersion}`);
        if (moduleExports === null || moduleExports === void 0 ? void 0 : moduleExports.BedrockRuntimeClient) {
            this._unwrap(moduleExports.BedrockRuntimeClient.prototype, "send");
            _isBedrockPatched = false;
        }
    }
    /**
     * Handles instrumentation for synchronous InvokeModel commands
     * Creates a span, extracts request attributes, executes the command, and processes the response
     *
     * @param command The InvokeModelCommand to instrument
     * @param original The original send method from the Bedrock client
     * @param client The Bedrock client instance
     * @returns {Promise<InvokeModelResponse>} The response from the InvokeModel call
     */
    handleInvokeModelCommand(args, command, original, client) {
        const span = this.oiTracer.startSpan("bedrock.invoke_model", {
            kind: api_1.SpanKind.INTERNAL,
        });
        const modelId = command.input.modelId;
        const system = (0, attribute_helpers_1.getSystemFromModelId)(modelId !== null && modelId !== void 0 ? modelId : "");
        (0, attribute_helpers_1.setBasicSpanAttributes)(span, system);
        const contextAttributes = (0, openinference_core_1.getAttributesFromContext)(api_1.context.active());
        span.setAttributes(contextAttributes);
        (0, invoke_model_request_attributes_1.extractInvokeModelRequestAttributes)({ span, command, system });
        try {
            const result = original.apply(client, args);
            // AWS SDK v3 send() method should always return a Promise
            if (!(0, types_1.isPromise)(result)) {
                api_1.diag.warn("Expected Promise from AWS SDK send method, got:", typeof result);
                span.setStatus({
                    code: api_1.SpanStatusCode.ERROR,
                    message: "Unexpected return type from AWS SDK",
                });
                span.end();
                return result;
            }
            return result
                .then((response) => {
                (0, invoke_model_response_attributes_1.extractInvokeModelResponseAttributes)({
                    span,
                    response,
                    modelType: system,
                });
                span.setStatus({ code: api_1.SpanStatusCode.OK });
                span.end();
                return response;
            })
                .catch((error) => {
                span.recordException(error);
                span.setStatus({
                    code: api_1.SpanStatusCode.ERROR,
                    message: error.message,
                });
                span.end();
                throw error;
            });
        }
        catch (error) {
            // Handle errors that occur before the Promise is returned (e.g. invalid parameters)
            if (error instanceof Error) {
                span.recordException(error);
                span.setStatus({ code: api_1.SpanStatusCode.ERROR, message: error.message });
            }
            else {
                span.setStatus({ code: api_1.SpanStatusCode.ERROR, message: String(error) });
            }
            span.end();
            throw error;
        }
    }
    /**
     * Handles instrumentation for streaming InvokeModel commands with guaranteed stream preservation
     *
     * This method ensures the original user stream is preserved regardless of instrumentation
     * success or failure. The user stream is returned immediately while instrumentation
     * happens asynchronously in the background using stream splitting.
     *
     * @param command The InvokeModelWithResponseStreamCommand to instrument
     * @param original The original send method from the Bedrock client
     * @param client The Bedrock client instance
     * @returns {Promise<{body: AsyncIterable<unknown>}>} Promise resolving to the response with preserved user stream
     */
    handleInvokeModelWithResponseStreamCommand(args, command, original, client) {
        const span = this.oiTracer.startSpan("bedrock.invoke_model", {
            kind: api_1.SpanKind.INTERNAL,
        });
        const modelId = command.input.modelId;
        const system = (0, attribute_helpers_1.getSystemFromModelId)(modelId !== null && modelId !== void 0 ? modelId : "");
        (0, attribute_helpers_1.setBasicSpanAttributes)(span, system);
        const contextAttributes = (0, openinference_core_1.getAttributesFromContext)(api_1.context.active());
        span.setAttributes(contextAttributes);
        (0, invoke_model_request_attributes_1.extractInvokeModelRequestAttributes)({
            span,
            command: command,
            system,
        });
        // Execute AWS SDK call and handle stream splitting outside error boundaries
        // This ensures the user stream is ALWAYS returned, regardless of instrumentation failures
        const result = original.apply(client, args);
        // Validate that we got a Promise as expected
        if (!(0, types_1.isPromise)(result)) {
            api_1.diag.warn("Expected Promise from AWS SDK send method, got:", typeof result);
            span.setStatus({
                code: api_1.SpanStatusCode.ERROR,
                message: "Unexpected return type from AWS SDK",
            });
            span.end();
            return result;
        }
        return result
            .then((response) => {
            // Guard against missing response body - return original response
            if (!response.body) {
                span.recordException(new Error("Response body is undefined"));
                span.setStatus({
                    code: api_1.SpanStatusCode.ERROR,
                    message: "Response body is undefined",
                });
                span.end();
                return response;
            }
            const { instrumentationStream, userStream } = (0, invoke_model_streaming_response_attributes_1.safelySplitStream)({
                originalStream: response.body,
            });
            // Start background instrumentation processing (non-blocking)
            // Only the instrumentation is wrapped in error handling - the user stream is protected
            if (instrumentationStream) {
                const consumed = (0, invoke_model_streaming_response_attributes_1.consumeBedrockStreamChunks)({
                    stream: instrumentationStream,
                    span,
                    modelType: system,
                });
                if (consumed == null) {
                    span.setStatus({ code: api_1.SpanStatusCode.OK });
                    span.end();
                }
                else {
                    consumed
                        .then(() => {
                        span.setStatus({ code: api_1.SpanStatusCode.OK });
                        span.end();
                    })
                        .catch((error) => {
                        span.recordException(error);
                        span.setStatus({
                            code: api_1.SpanStatusCode.ERROR,
                            message: error.message,
                        });
                        span.end();
                    });
                }
            }
            else {
                api_1.diag.debug("No instrumentation stream available, ending span cleanly");
                span.setStatus({ code: api_1.SpanStatusCode.OK });
                span.end();
            }
            // Return user stream immediately - instrumentation cannot interfere
            return Object.assign(Object.assign({}, response), { body: userStream });
        })
            .catch((error) => {
            // If the AWS SDK call itself fails, record error but still try to return original response
            span.recordException(error);
            span.setStatus({
                code: api_1.SpanStatusCode.ERROR,
                message: error.message,
            });
            span.end();
            throw error; // Re-throw since we have no stream to return
        });
    }
    /**
     * Handles instrumentation for Converse API commands
     * Creates a span, extracts request attributes, executes the command, and processes the response
     *
     * @param command The ConverseCommand to instrument
     * @param original The original send method from the Bedrock client
     * @param client The Bedrock client instance
     * @returns {Promise<ConverseResponse>} The response from the Converse call
     */
    handleConverseCommand(args, command, original, client) {
        const span = this.oiTracer.startSpan("bedrock.converse", {
            kind: api_1.SpanKind.INTERNAL,
        });
        const modelId = command.input.modelId;
        const system = (0, attribute_helpers_1.getSystemFromModelId)(modelId !== null && modelId !== void 0 ? modelId : "");
        (0, attribute_helpers_1.setBasicSpanAttributes)(span, system);
        const contextAttributes = (0, openinference_core_1.getAttributesFromContext)(api_1.context.active());
        span.setAttributes(contextAttributes);
        (0, converse_request_attributes_1.extractConverseRequestAttributes)({ span, command });
        try {
            const result = original.apply(client, args);
            // AWS SDK v3 send() method should always return a Promise
            if (!(0, types_1.isPromise)(result)) {
                api_1.diag.warn("Expected Promise from AWS SDK send method, got:", typeof result);
                span.setStatus({
                    code: api_1.SpanStatusCode.ERROR,
                    message: "Unexpected return type from AWS SDK",
                });
                span.end();
                return result;
            }
            return result
                .then((response) => {
                (0, converse_response_attributes_1.extractConverseResponseAttributes)({ span, response });
                span.setStatus({ code: api_1.SpanStatusCode.OK });
                span.end();
                return response;
            })
                .catch((error) => {
                span.recordException(error);
                span.setStatus({
                    code: api_1.SpanStatusCode.ERROR,
                    message: error.message,
                });
                span.end();
                throw error;
            });
        }
        catch (error) {
            // Handle errors that occur before the Promise is returned (e.g. invalid parameters)
            if (error instanceof Error) {
                span.recordException(error);
                span.setStatus({ code: api_1.SpanStatusCode.ERROR, message: error.message });
            }
            else {
                span.setStatus({ code: api_1.SpanStatusCode.ERROR, message: String(error) });
            }
            span.end();
            throw error;
        }
    }
    /**
     * Handles instrumentation for streaming Converse commands with guaranteed stream preservation
     *
     * This method ensures the original user stream is preserved regardless of instrumentation
     * success or failure. The user stream is returned immediately while instrumentation
     * happens asynchronously in the background using stream splitting.
     *
     * @param command The ConverseStreamCommand to instrument
     * @param original The original send method from the Bedrock client
     * @param client The Bedrock client instance
     * @returns {Promise<{stream: AsyncIterable<unknown>}>} Promise resolving to the response with preserved user stream
     */
    handleConverseStreamCommand(args, command, original, client) {
        const span = this.oiTracer.startSpan("bedrock.converse", {
            kind: api_1.SpanKind.INTERNAL,
        });
        const modelId = command.input.modelId;
        const system = (0, attribute_helpers_1.getSystemFromModelId)(modelId !== null && modelId !== void 0 ? modelId : "");
        (0, attribute_helpers_1.setBasicSpanAttributes)(span, system);
        const contextAttributes = (0, openinference_core_1.getAttributesFromContext)(api_1.context.active());
        span.setAttributes(contextAttributes);
        (0, converse_request_attributes_1.extractConverseRequestAttributes)({
            span,
            command: command,
        });
        // Execute AWS SDK call and handle stream splitting outside error boundaries
        // This ensures the user stream is ALWAYS returned, regardless of instrumentation failures
        const result = original.apply(client, args);
        // Validate that we got a Promise as expected
        if (!(0, types_1.isPromise)(result)) {
            api_1.diag.warn("Expected Promise from AWS SDK send method, got:", typeof result);
            span.setStatus({
                code: api_1.SpanStatusCode.ERROR,
                message: "Unexpected return type from AWS SDK",
            });
            span.end();
            return result;
        }
        return result
            .then((response) => {
            // Guard against missing response stream - return original response
            if (!response.stream) {
                span.recordException(new Error("Response stream is undefined"));
                span.setStatus({
                    code: api_1.SpanStatusCode.ERROR,
                    message: "Response stream is undefined",
                });
                span.end();
                return response;
            }
            const { instrumentationStream, userStream } = (0, invoke_model_streaming_response_attributes_1.safelySplitStream)({
                originalStream: response.stream,
            });
            // Start background instrumentation processing (non-blocking)
            // Only the instrumentation is wrapped in error handling - the user stream is protected
            if (instrumentationStream) {
                // @ts-expect-error - instrumentationStream is guaranteed non-null by the if check above
                (0, converse_streaming_response_attributes_1.consumeConverseStreamChunks)({
                    stream: instrumentationStream,
                    span,
                })
                    .then(() => {
                    span.setStatus({ code: api_1.SpanStatusCode.OK });
                    span.end();
                })
                    .catch((error) => {
                    span.recordException(error);
                    span.setStatus({
                        code: api_1.SpanStatusCode.ERROR,
                        message: error.message,
                    });
                    span.end();
                });
            }
            else {
                api_1.diag.debug("No instrumentation stream available, ending span cleanly");
                span.setStatus({ code: api_1.SpanStatusCode.OK });
                span.end();
            }
            // Return user stream immediately - instrumentation cannot interfere
            return Object.assign(Object.assign({}, response), { stream: userStream });
        })
            .catch((error) => {
            // If the AWS SDK call itself fails, record error but still try to return original response
            span.recordException(error);
            span.setStatus({
                code: api_1.SpanStatusCode.ERROR,
                message: error.message,
            });
            span.end();
            throw error; // Re-throw since we have no stream to return
        });
    }
}
exports.BedrockInstrumentation = BedrockInstrumentation;
