// @arizeai/openinference-core/dist/src/trace/trace-config/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultTraceConfig = exports.traceConfigMetadata = exports.REDACTED_VALUE = exports.DEFAULT_HIDE_PROMPTS = exports.DEFAULT_BASE64_IMAGE_MAX_LENGTH = exports.DEFAULT_HIDE_EMBEDDING_VECTORS = exports.DEFAULT_HIDE_OUTPUT_TEXT = exports.DEFAULT_HIDE_INPUT_TEXT = exports.DEFAULT_HIDE_INPUT_IMAGES = exports.DEFAULT_HIDE_OUTPUT_MESSAGES = exports.DEFAULT_HIDE_INPUT_MESSAGES = exports.DEFAULT_HIDE_OUTPUTS = exports.DEFAULT_HIDE_INPUTS = exports.DEFAULT_HIDE_LLM_TOOLS = exports.OPENINFERENCE_HIDE_PROMPTS = exports.OPENINFERENCE_BASE64_IMAGE_MAX_LENGTH = exports.OPENINFERENCE_HIDE_EMBEDDING_VECTORS = exports.OPENINFERENCE_HIDE_OUTPUT_TEXT = exports.OPENINFERENCE_HIDE_INPUT_TEXT = exports.OPENINFERENCE_HIDE_INPUT_IMAGES = exports.OPENINFERENCE_HIDE_OUTPUT_MESSAGES = exports.OPENINFERENCE_HIDE_INPUT_MESSAGES = exports.OPENINFERENCE_HIDE_OUTPUTS = exports.OPENINFERENCE_HIDE_INPUTS = exports.OPENINFERENCE_HIDE_LLM_TOOLS = void 0;
/** Hides the tool definitions advertised to the LLM */
exports.OPENINFERENCE_HIDE_LLM_TOOLS = "OPENINFERENCE_HIDE_LLM_TOOLS";
/** Hides input value & messages */
exports.OPENINFERENCE_HIDE_INPUTS = "OPENINFERENCE_HIDE_INPUTS";
/** Hides output value & messages */
exports.OPENINFERENCE_HIDE_OUTPUTS = "OPENINFERENCE_HIDE_OUTPUTS";
/** Hides all input messages */
exports.OPENINFERENCE_HIDE_INPUT_MESSAGES = "OPENINFERENCE_HIDE_INPUT_MESSAGES";
/** Hides all output messages */
exports.OPENINFERENCE_HIDE_OUTPUT_MESSAGES = "OPENINFERENCE_HIDE_OUTPUT_MESSAGES";
/** Hides images from input messages */
exports.OPENINFERENCE_HIDE_INPUT_IMAGES = "OPENINFERENCE_HIDE_INPUT_IMAGES";
/** Hides text from input messages */
exports.OPENINFERENCE_HIDE_INPUT_TEXT = "OPENINFERENCE_HIDE_INPUT_TEXT";
/** Hides text from output messages */
exports.OPENINFERENCE_HIDE_OUTPUT_TEXT = "OPENINFERENCE_HIDE_OUTPUT_TEXT";
/** Hides embedding vectors */
exports.OPENINFERENCE_HIDE_EMBEDDING_VECTORS = "OPENINFERENCE_HIDE_EMBEDDING_VECTORS";
/** Limits characters of a base64 encoding of an image */
exports.OPENINFERENCE_BASE64_IMAGE_MAX_LENGTH = "OPENINFERENCE_BASE64_IMAGE_MAX_LENGTH";
/** Hides LLM prompts */
exports.OPENINFERENCE_HIDE_PROMPTS = "OPENINFERENCE_HIDE_PROMPTS";
exports.DEFAULT_HIDE_LLM_TOOLS = false;
exports.DEFAULT_HIDE_INPUTS = false;
exports.DEFAULT_HIDE_OUTPUTS = false;
exports.DEFAULT_HIDE_INPUT_MESSAGES = false;
exports.DEFAULT_HIDE_OUTPUT_MESSAGES = false;
exports.DEFAULT_HIDE_INPUT_IMAGES = false;
exports.DEFAULT_HIDE_INPUT_TEXT = false;
exports.DEFAULT_HIDE_OUTPUT_TEXT = false;
exports.DEFAULT_HIDE_EMBEDDING_VECTORS = false;
exports.DEFAULT_BASE64_IMAGE_MAX_LENGTH = 32000;
exports.DEFAULT_HIDE_PROMPTS = false;
/** When a value is hidden, it will be replaced by this redacted value */
exports.REDACTED_VALUE = "__REDACTED__";
/**
 * The default, environment, and type information for each value on the TraceConfig
 * Used to generate a full TraceConfig object with the correct types and default values
 */
exports.traceConfigMetadata = {
    hideLLMTools: {
        default: exports.DEFAULT_HIDE_LLM_TOOLS,
        envKey: exports.OPENINFERENCE_HIDE_LLM_TOOLS,
        type: "boolean",
    },
    hideInputs: {
        default: exports.DEFAULT_HIDE_INPUTS,
        envKey: exports.OPENINFERENCE_HIDE_INPUTS,
        type: "boolean",
    },
    hideOutputs: {
        default: exports.DEFAULT_HIDE_OUTPUTS,
        envKey: exports.OPENINFERENCE_HIDE_OUTPUTS,
        type: "boolean",
    },
    hideInputMessages: {
        default: exports.DEFAULT_HIDE_INPUT_MESSAGES,
        envKey: exports.OPENINFERENCE_HIDE_INPUT_MESSAGES,
        type: "boolean",
    },
    hideOutputMessages: {
        default: exports.DEFAULT_HIDE_OUTPUT_MESSAGES,
        envKey: exports.OPENINFERENCE_HIDE_OUTPUT_MESSAGES,
        type: "boolean",
    },
    hideInputImages: {
        default: exports.DEFAULT_HIDE_INPUT_IMAGES,
        envKey: exports.OPENINFERENCE_HIDE_INPUT_IMAGES,
        type: "boolean",
    },
    hideInputText: {
        default: exports.DEFAULT_HIDE_INPUT_TEXT,
        envKey: exports.OPENINFERENCE_HIDE_INPUT_TEXT,
        type: "boolean",
    },
    hideOutputText: {
        default: exports.DEFAULT_HIDE_OUTPUT_TEXT,
        envKey: exports.OPENINFERENCE_HIDE_OUTPUT_TEXT,
        type: "boolean",
    },
    hideEmbeddingVectors: {
        default: exports.DEFAULT_HIDE_EMBEDDING_VECTORS,
        envKey: exports.OPENINFERENCE_HIDE_EMBEDDING_VECTORS,
        type: "boolean",
    },
    base64ImageMaxLength: {
        default: exports.DEFAULT_BASE64_IMAGE_MAX_LENGTH,
        envKey: exports.OPENINFERENCE_BASE64_IMAGE_MAX_LENGTH,
        type: "number",
    },
    hidePrompts: {
        default: exports.DEFAULT_HIDE_PROMPTS,
        envKey: exports.OPENINFERENCE_HIDE_PROMPTS,
        type: "boolean",
    },
};
exports.DefaultTraceConfig = {
    hideLLMTools: exports.DEFAULT_HIDE_LLM_TOOLS,
    hideInputs: exports.DEFAULT_HIDE_INPUTS,
    hideOutputs: exports.DEFAULT_HIDE_OUTPUTS,
    hideInputMessages: exports.DEFAULT_HIDE_INPUT_MESSAGES,
    hideOutputMessages: exports.DEFAULT_HIDE_OUTPUT_MESSAGES,
    hideInputImages: exports.DEFAULT_HIDE_INPUT_IMAGES,
    hideInputText: exports.DEFAULT_HIDE_INPUT_TEXT,
    hideOutputText: exports.DEFAULT_HIDE_OUTPUT_TEXT,
    hideEmbeddingVectors: exports.DEFAULT_HIDE_EMBEDDING_VECTORS,
    base64ImageMaxLength: exports.DEFAULT_BASE64_IMAGE_MAX_LENGTH,
    hidePrompts: exports.DEFAULT_HIDE_PROMPTS,
};
